Complete the whole cycle end-to-end: implement merging of video and check in the test

neomatrix369 opened the pull request:
- implement merging logic in download_and_merge_video.sh
- upload merged video into the S3 bucket
- send the SNS event via queue-client
- download merged video from S3 bucket
- compare it with expected merged video (WIP: comparing videos - test is failing at this point)

@julianghionoiu I think you did mention about the issue with comparing two videos event if they are the same, what are your thoughts here?

neomatrix369 committed:
Complete the whole cycle end-to-end:
- implement merging logic in download_and_merge_video.sh
- upload merged video into the S3 bucket
- send the SNS event via queue-client
- download merged video from S3 bucket
- compare it with expected merged video (WIP: comparing videos - test is failing at this point)

neomatrix369 commented on src/test/java/tdl/datapoint/video/VideoDatapointAcceptanceTest.java:
```diff
@@ -117,24 +119,48 @@ public void create_repo_and_upload_video() throws Exception {
+        assertThat("participantId matching", rawVideoUploaded.getParticipant(), equalTo(participantId));
+        assertThat("challengeId matching", rawVideoUploaded.getChallengeId(), equalTo(challengeId));
+        String key = String.format("%s/%s/real-recording.mp4", challengeId, participantId);
+        assertThat("Video link match check: expecting something like s3://tdl-official-videos/.../real-recording.mp4", rawVideoUploaded.getVideoLink(), equalTo(
```
Im using the same bucket name as in production otherwise test does not find the merged video

neomatrix369 commented on src/test/java/tdl/datapoint/video/VideoDatapointAcceptanceTest.java:
```diff
@@ -117,24 +119,48 @@ public void create_repo_and_upload_video() throws Exception {
+        assertThatFilesAreEqual("Expect the files to match in content", actualAccumulatorVideo, expectedAccumulatorVideo);
+    }
+
+    private void assertThatFilesAreEqual(String reason, Path firstFile, Path secondFile) throws IOException {
```
How to compare the videos using some video related library?

neomatrix369 replied:
Maybe `humble-video-arch-x86_64-...`?

neomatrix369 committed:
Applying first set of changes as per discussions during code review session with Julian - test is still failing

neomatrix369 committed:
Upload first video when there is no accumulator video present at all and compare the actual results with the expected results

neomatrix369 committed:
Upload second video when there is no accumulator video present at all and compare the actual results with the expected results

neomatrix369 committed:
Use two buckets instead of one, move the split videos into the `tdl-official-split-videos` bucket and the accumulated videos into the `tdl-official-videos` bucket, adapt implementation and test code to support it

neomatrix369 committed:
Apply dual bucket (split & accumulated) config to cloudfoundry, serverless config files and s3-sns event and README.md

neomatrix369 commented:
I have applied the below with the new commits:
- split buckets into two (split and accumulated)
- split test into first video upload and second video uploads
- applied the dual bucket config across cloudfoundry and serverless config (this needs extra checking)

The splitting bucket should help with conflicting events or incorrect triggering of events when accumulated video is uploaded.

Pending:
- generate http based URL to download video from S3 bucket
- comparing videos for both tests

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoDatapointAcceptanceTest.java:
```diff
@@ -113,28 +122,89 @@ public void tearDown() throws Exception {
+        String challengeId = "TCH";
+        String participantId = generateId();
+        String s3AccumulatorVideoDestination = String.format("%s/%s/real-recording.mp4", challengeId, participantId);
+        TestVideoFile accumulatorVideo = new TestVideoFile("tdl/datapoint/video/second_video_upload/before/real-recording.mp4");
```
The given should place the accumulator into Minio.

neomatrix369 replied:
How is that done via the test? I think the S3.putObject command would do that?

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoDatapointAcceptanceTest.java:
```diff
@@ -113,28 +122,89 @@ public void tearDown() throws Exception {
+    }
+
+    @Test
+    public void upload_second_video() throws Exception {
```
The names of this test should be: upload_when_accumulator_video_exists

julianghionoiu commented on container/images/base/download_and_merge_video.sh:
```diff
@@ -26,13 +29,39 @@ ensure_env "SQS_QUEUE_URL"
+rm ${VIDEOS_LIST} || true
+if [[ -s "${TARGET_VIDEO_NAME}" ]]; then
+   echo "file '${TARGET_VIDEO_NAME}'" > ${VIDEOS_LIST}
+   ls screencast_* -1 | xargs -n1 -I {} echo "file '{}'" >> ${VIDEOS_LIST}
```
The list should explicitly be accumulator, followed by new video. Otherwise we might get them in the wrong order.

neomatrix369 replied:
I believed It is in that order:
```
echo "file '${TARGET_VIDEO_NAME}'" > ${VIDEOS_LIST}  <=== accumulator video
```
```
ls screencast_* -1 | xargs -n1 -I {} echo "file '{}'" >> ${VIDEOS_LIST}   <=== split video to be merged to the bottom of the accumulator
```

julianghionoiu replied:
We are not going to loop through the received files

julianghionoiu commented on ecs-cluster-definition/templates/cf-master.jinja2.json:
```diff
@@ -41,14 +41,29 @@
                                         "s3:GetObject"
                                     ],
-                                    "Resource": "{{ S3_BUCKET_ARN }}/*"
+                                    "Resource": "{{ S3_ACCUMULATED_VIDEOS_BUCKET_ARN }}/*"
```
Because the accumulated videos bucket is only used by this service, it's definition should be managed by this CloudFormation script. 
We would need to add when we trigger the first deployment.

julianghionoiu commented on src/main/java/tdl/datapoint/video/VideoUploadHandler.java:
```diff
@@ -103,10 +103,22 @@ public String handleRequest(Map<String, Object> s3EventMap, Context context) {
         String challengeId = event.getChallengeId();
 
+        final String ACCUMULATOR_VIDEO_NAME = "real-recording.mp4";
+        final String SPLIT_VIDEOS_BUCKET_NAME = "tdl-official-split-videos";
```
The name of the bucket is irrelevant for the code so should be injected from the outside.

neomatrix369 replied:
Will extract eventually so to make VideoUploadHandler more generic

neomatrix369 replied:
Done!

julianghionoiu commented on src/main/java/tdl/datapoint/video/VideoUploadHandler.java:
```diff
@@ -103,10 +103,22 @@ public String handleRequest(Map<String, Object> s3EventMap, Context context) {
+
+        final String s3UrlNewVideo = String.format("s3://%s/%s", SPLIT_VIDEOS_BUCKET_NAME, event.getKey());
+
+        final String s3BucketKey = String.format("%s/%s/%s", challengeId, participantId, ACCUMULATOR_VIDEO_NAME);
```
As discussed, the accumulator key should follow a naming convention that will allow us to make the bucket public. 
Should be:
```
challengeId/participantId/hash(challengeId, participantId, secret)/recording.mp4
```

julianghionoiu replied:
The secret is passed as a config value on deployment

neomatrix369 replied:
Done!

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoDatapointAcceptanceTest.java:
```diff
@@ -113,28 +122,89 @@ public void tearDown() throws Exception {
+        Path expectedAccumulatorVideo = new TestVideoFile("tdl/datapoint/video/second_video_upload/after/real-recording.mp4").asFile().toPath();
+        //TODO use Http instead of s3
+        Path actualAccumulatorVideo = new TestVideoFile(localS3AccumulatedVideoBucket, key).getS3Object("real-recording", ".mp4").toPath();
+        //TODO compare videos using dev-screen-record's logic: see acceptance test that generates and reads QRcode
```
I have managed to locate the piece of code I was telling you about:
- Videos with barcode: https://github.com/julianghionoiu/dev-screen-anonymise/blob/master/src/test/resources/rec_barcode_matrix/recording_from_matrix_of_barcodes_at_4x.mp4
- how to generate one: https://github.com/julianghionoiu/dev-screen-anonymise/blob/master/src/test/java/tdl/record/image/input/GenerateInputWithMatrixOfBarcodes.java
- how to read barcodes: https://github.com/julianghionoiu/dev-screen-anonymise/blob/master/src/test/resources/rec_barcode_matrix/recording_from_matrix_of_barcodes_at_4x.mp4

neomatrix369 replied:
Will look into it asap

julianghionoiu replied:
In our case, that needs to be a one barcode recording. The functionality needs to be extracted in a separate test library anyway

neomatrix369 committed:
Renamed tests as per code review comment from @JulianG

neomatrix369 committed:
Moved the step to place accumulator into the S3 bucket to the Given step

neomatrix369 committed:
Publishing the shareable S3 url (http-based) after merging the video, although the = to signs in the URL itself is getting truncated due to some logic in queue-client

neomatrix369 committed:
Replacing logic to get a list of screencast files with the value in NEW_VIDEO, the new video to be merged

neomatrix369 committed:
Generate downloadable file URL, also add a note in the README.md to make tdl-official-videos publicly readable

We are not presigining the URL, just constructing one that works as per AWS conventions

neomatrix369 committed:
Extracting out constants/hard-coded values from VideoUploadHandler.java
Injecting S3 url for  new Video, accumulator video and accumulator video filename into ECS task and docker container from the tests and the VideoUploadHandler.java

neomatrix369 committed:
Create a unique hash from participantId, challengeId and some secret for the publically available accumulator video file

neomatrix369 committed:
Moved common Given clauses to the setup() section of the acceptance test, removing some duplications in use of constants

neomatrix369 commented on src/test/java/tdl/datapoint/video/VideoDatapointAcceptanceTest.java:
```diff
@@ -40,6 +41,12 @@
     private static final int TASK_FINISH_CHECK_RETRY_COUNT = 10;
+    private static final int FIRST_ACCUMULATED_VIDEO_EVENT = 0;
+
+    private static final String TDL_OFFICIAL_SPLIT_VIDEOS = "tdl-official-split-videos";
```
I can move these four into the `local.params.yml` and `dev.params.yml` file

neomatrix369 commented:
@julianghionoiu in https://github.com/julianghionoiu/dpnt-video/pull/7#discussion_r214547376 wrt to how to read barcodes, did you mean this class https://github.com/julianghionoiu/dev-screen-anonymise/blob/master/src/test/java/tdl/record/image/output/OutputToBarcodeMatrixReader.java

julianghionoiu commented:
Yes, but that example is more complex as it uses a matrix of barcodes.
On Wed, 5 Sep 2018 at 13:13, mani <notifications@github.com> wrote:

> @julianghionoiu <https://github.com/julianghionoiu> in #7 (comment)
> <https://github.com/julianghionoiu/dpnt-video/pull/7#discussion_r214547376>
> wrt to how to read barcodes, did you mean this class
> https://github.com/julianghionoiu/dev-screen-anonymise/blob/master/src/test/java/tdl/record/image/output/OutputToBarcodeMatrixReader.java
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/julianghionoiu/dpnt-video/pull/7#issuecomment-418707895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2qKYOiCe8baOYDuNuMiGmO7zOlWwreks5uX7_QgaJpZM4WUYuk>
> .
>

julianghionoiu commented:
In fact I am ok with using the matrix of barcodes, even if it is bit
overkill. If we can pull that into a separate library, it would be great.


On Wed, 5 Sep 2018 at 13:35, Iulian Ghionoiu <iuli.constantin.ro@gmail.com>
wrote:

> Yes, but that example is more complex as it uses a matrix of barcodes.
> On Wed, 5 Sep 2018 at 13:13, mani <notifications@github.com> wrote:
>
>> @julianghionoiu <https://github.com/julianghionoiu> in #7 (comment)
>> <https://github.com/julianghionoiu/dpnt-video/pull/7#discussion_r214547376>
>> wrt to how to read barcodes, did you mean this class
>> https://github.com/julianghionoiu/dev-screen-anonymise/blob/master/src/test/java/tdl/record/image/output/OutputToBarcodeMatrixReader.java
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/julianghionoiu/dpnt-video/pull/7#issuecomment-418707895>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AE2qKYOiCe8baOYDuNuMiGmO7zOlWwreks5uX7_QgaJpZM4WUYuk>
>> .
>>
>

neomatrix369 commented:
I'm looking at it and I think with what you have I can write a simple video comparer and make it work with our bar/qr code videos and then extract it away when it all works

neomatrix369 commented:
I agree about the complexity, partly because of the layout of the project folders, I'll try to simplify it

neomatrix369 committed:
Replaced provided example videos with barcode and qrcode videos, this should not change the logic of the implementation and tests so far

neomatrix369 committed:
Adding scripts to start and stop the local dependencies (AWS test stubs) for the tests to work

neomatrix369 committed:
Compare two videos using classes that compare barcode and qrcodes in the videos frames, rendering the end-to-end test to pass

neomatrix369 committed:
Extract the parts to do with comparing the video files and asserting on the video events and its properties into a class of its own

neomatrix369 committed:
Refactoring test class VideoMergingAcceptanceTest, reorganising methods and minor renamings

neomatrix369 committed:
Updated the README.md with info on the start and stop scripts to run the local AWS simulators

neomatrix369 committed:
Changed the coupling between raw video events and CompareVideos, moved it from constructing during construction to when assertion on the videos are done

neomatrix369 committed:
Refactoring startExternalDependencies.sh for better readability

neomatrix369 committed:
Refactoring stopExternalDependencies.sh for better readability

julianghionoiu commented on src/main/java/tdl/datapoint/video/VideoUploadHandler.java:
```diff
@@ -34,9 +37,21 @@
 
     @SuppressWarnings("WeakerAccess")
-    public VideoUploadHandler() {
+    public VideoUploadHandler(String accumulatorVideoName,
```
The video filename cannot be in the handler constructor.
The way a lambda works is:
- handler is constructed on startup
- handler is invoked via `handleRequest` and will be reused for multiple requests

julianghionoiu commented on src/main/java/tdl/datapoint/video/VideoUploadHandler.java:
```diff
@@ -34,9 +37,21 @@
     @SuppressWarnings("WeakerAccess")
-    public VideoUploadHandler() {
+    public VideoUploadHandler(String accumulatorVideoName,
+                              String splitVideosBucketName,
```
This bucket name is already part of the SNS event and should be read from there

neomatrix369 replied:
This is now resolved, is it okay for the remaining params to be passed via the constructor of `VideoUploadHandler`?

julianghionoiu commented on src/main/java/tdl/datapoint/video/VideoUploadHandler.java:
```diff
@@ -93,20 +108,61 @@ private static AmazonECS createECSClient(String serviceEndpoint, String signingR
+
+    public String createHashFrom(String challengeId, String participantId, String secret) {
+        try {
+            return makeSHA1Hash(challengeId + participantId + secret);
```
Use `HMAC`

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoMergingAcceptanceTest.java:
```diff
@@ -49,92 +56,127 @@
+    public void upload_first_video_when_accumulator_does_not_exist_yet() throws Exception {
+        // Given - The participant produces Video files while solving a challenge
+        TestVideoFile accumulatorVideo = new TestVideoFile("tdl/datapoint/video/first_video_upload/before/" + ACCUMULATOR_VIDEO_FILENAME);
+        localS3AccumulatedVideoBucket.putObject(accumulatorVideo.asFile(), s3AccumulatorVideoDestination);
```
Accumulator should be empty here. The given should actually ensure that it is not present.

neomatrix369 replied:
Accumulator is empty, its zero bytes, I can check for this and ensure it is so.

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoMergingAcceptanceTest.java:
```diff
@@ -49,92 +56,127 @@
+        // Given - The participant produces Video files while solving a challenge
+        TestVideoFile accumulatorVideo = new TestVideoFile("tdl/datapoint/video/first_video_upload/before/" + ACCUMULATOR_VIDEO_FILENAME);
+        localS3AccumulatedVideoBucket.putObject(accumulatorVideo.asFile(), s3AccumulatorVideoDestination);
+        String s3destination = String.format("%s/%s/screencast_1.mp4", challengeId, participantId);
```
In order for the tests not to interact, we should use a different participantId, ideally it should be unique for each test run.

neomatrix369 replied:
I think I got mislead with this comment, how is `participantId` not unique per test? The @Before is run before each test case, so unique participantIds are generated, i;ll revert my changes with regard to this specific comment

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoMergingAcceptanceTest.java:
```diff
@@ -49,92 +56,127 @@
+        TestVideoFile accumulatorVideo = new TestVideoFile("tdl/datapoint/video/first_video_upload/before/" + ACCUMULATOR_VIDEO_FILENAME);
+        localS3AccumulatedVideoBucket.putObject(accumulatorVideo.asFile(), s3AccumulatorVideoDestination);
+        String s3destination = String.format("%s/%s/screencast_1.mp4", challengeId, participantId);
+        TestVideoFile newVideo = new TestVideoFile("screencast_20180727T144854.mp4");
```
Make this part of the same resource folder "first_video_upload"

neomatrix369 replied:
Done!

julianghionoiu commented on src/test/java/tdl/datapoint/video/barcode/OutputToBarcodeMatrixReader.java:
```diff
@@ -0,0 +1,128 @@
+        try {
+            int subWidth = image.getWidth()/2;
+            int subHeight = image.getHeight()/2;
+            String topLeftPayload = decodeBarcode(
```
This can be simplified to only render one barcode, not four.

neomatrix369 replied:
Im thinking this change might tangent us from the real purpose of the PR, we could tackle the content of the video and the comparison classes as a separate concern to this PR or the purpose of this project

neomatrix369 committed:
Individual videos (.mp4) files have been made part of the respective resource folders: first_xxxx and second_xxxx folders

neomatrix369 committed:
Replace all the unescaped characters in the signed token with 0, to prevent the URL from being incorrectly parsed by the key-value tokeniser

julianghionoiu commented on src/test/java/tdl/datapoint/video/VideoMergingAcceptanceTest.java:
```diff
@@ -152,6 +148,15 @@ public void upload_when_accumulator_video_exists() throws Exception {
 
     //~~~~~~~~~~ Helpers ~~~~~~~~~~~~~`
 
+    private String generateUniqueParticipant() {
```
This method does more than it says. I was not expecting the comapreVideos and accumulator destination to be instantiated here.

neomatrix369 replied:
Reverted changes

julianghionoiu commented on src/main/java/tdl/datapoint/video/security/TokenSigningService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Arrays;
+import java.util.Base64;
+
+public class TokenSigningService {
```
It would be good to have a unit test for this class, so that we capture it's behaviour.

julianghionoiu replied:
The most important test is one that proves the idempotency of the hash. For the same inputs we should always get the same hash, otherwise we would not be able to locate the accumulator.

neomatrix369 replied:
Onto it now!

neomatrix369 replied:
Done!

julianghionoiu commented on src/main/java/tdl/datapoint/video/security/TokenSigningService.java:
```diff
@@ -0,0 +1,43 @@
+        return new SecretKeySpec(encryptionKeyAsBytes, 0, encryptionKeyAsBytes.length, "AES");
+    }
+
+    private String sign(String token) {
```
This is not signing as the message cannot be recovered.

neomatrix369 replied:
Will change that

neomatrix369 replied:
Done, we can discuss further if changes are needed

julianghionoiu reviewed (commented):
Nice

julianghionoiu commented on src/main/java/tdl/datapoint/video/security/TokenSigningService.java:
```diff
@@ -37,6 +37,15 @@ private String sign(String token) {
     public String createHashFrom(String challengeId, String participantId) throws UnsupportedEncodingException {
         String signedToken = sign(challengeId + participantId);
         String encodeToken = Base64.getEncoder().encodeToString(signedToken.getBytes("UTF-8"));
+        encodeToken = (String) encodeToken
```
We need a URL encoded string right? 
You can do something like:
```
import java.util.Base64;
return Base64.getUrlEncoder().encodeToString(bytes);
```

neomatrix369 replied:
This does not work, if you see what I'm doing which isn't the perfect solution but eliminates characters like `=` and others from tokens like this:
```
Wy0xMjgsIDUsIDEyMCwgLTksIDEzLCAtNTAsIDgxLCAtNzEsIDE1LCA4LCAtODMsIC0zMSwgMTIzLCAtMTA1LCAtNDQsIDc5XQ==
```

neomatrix369 replied:
There should be a better way to do this but for now the closest solution

neomatrix369 replied:
Not too happy with the solution atm, but lets discuss this further

neomatrix369 committed:
Renaming the token encryption class to more appropriate name, some changes to the encryption logic

neomatrix369 committed:
Reverting changes wrt unique participantId in VideoMergingAcceptanceTest.java

neomatrix369 committed:
Refactoring multiple replaceAll with regex string expression

neomatrix369 committed:
Minor refactoring of the bytesToHex) in TokenEncryptionService

neomatrix369 committed:
Removing redundant passing of split video bucket - already in SNS event

neomatrix369 committed:
Inject secret through environment variables, rather than via the constructor

neomatrix369 committed:
Minor inlining of variable in VideoUploadHandler

neomatrix369 committed:
Adding a check to first test (upload_first_video_when_accumulator_does_not_exist_yet) to see if the accumulator video exists or not

neomatrix369 committed:
Adding a check to second test (upload_when_accumulator_video_exists) to see if the accumulator video exists or not zero in size

neomatrix369 committed:
Added tests around TokenEncryptionServiceTest to ensure they generate expected encrypted tokens and also that the service is idempotent for the same given inputs

neomatrix369 commented:
@julianghionoiu Just wanted to comment that the below command:

```bash
./gradlew --rerun-tasks test  
```

only runs the `VideoMergingAcceptanceTest`

and in order to run a specific test i.e. `TokenEncryptionServiceTest` I had to do this:

```bash
./gradlew clean test --tests *TokenEncryptionServiceTest -i
```
